File: wildcard-matching/wildcard-matching.cpp
class Solution {
public:
    
    bool solver(string &s, string &p,int i,int j,vector<vector<int>> &dp)
    {
        if(i>=s.length() && j>=p.length() ) return true;
        if( j>=p.length() )
            return false;
        bool match=false;
        if(dp[i][j]!= -1)
            return dp[i][j];
        match = (i<s.length() and (s[i]==p[j] or p[j]=='?') );
        if(p[j]=='*')
        {   if(i<s.length())
                return dp[i][j] = solver(s,p,i+1,j,dp) or (solver(s,p,i,j+1,dp));
            else
                return dp[i][j] = (solver(s,p,i,j+1,dp));
        }
        if(match)
            return dp[i][j] = solver(s,p,i+1,j+1,dp);
        
        return false;
    }   
        
    bool isMatch(string s, string p) {
    vector<vector<int>> dp(s.length()+1,vector<int>(p.length()+1,-1));
    return solver(s,p,0,0,dp);
    }
};

File: wildcard-matching/wildcard-matching.py
class Solution(object):
    def isMatch(self, s, p):
        n = len(s)
        m = len(p)
        dp = [[False]*(m+1) for i in range(n+1)]
        dp[0][0] = True
        for j in range(1,m+1):
            if p[j-1] == '*':
                dp[0][j] = dp[0][j-1]
        
        for i in range(1,len(s)+1):
            for j in range(1,len(p)+1):
                dp[i][j] = (p[j-1] in [s[i-1],'?','*'] and dp[i-1][j-1]) or (p[j-1] == '*' and (dp[i][j-1] or dp[i-1][j]))
  
        return dp[n][m]
        